### PR TITLE
Support nativeTarget to build Scala Native static/shared libraries

### PR DESCRIPTION
**Pull request by keynmol**
Scala Native CLI supports building dynamic and shared libraries using `--build-target` argument.

So far it hasn't been supported. This PR adds the support, by

- Adding nativeTarget directive and native-target CLI option
- Splitting PackageType.Native into 3 different types
- Modifying the logic to not throw when main class is not found if we are building a library
- **IMPORTANT**: use `-Wl,install_name` argument to rename the library inside its own metadata, otherwise it points to `.scala-build/project_asdkljadljk190/main` - instead of `test.dylib` that we want

### CODE REVIEW

File: modules/options/src/main/scala/scala/build/options/ScalaNativeOptions.scala
Comment (lolgab) [inline modules/options/src/main/scala/scala/build/options/ScalaNativeOptions.scala:57]: Have you considered `dynamic-library` and `static-library`?

Comment (keynmol) [reply modules/options/src/main/scala/scala/build/options/ScalaNativeOptions.scala:57]: These values actually have to match what Scala Native CLI uses: https://github.com/scala-native/scala-native-cli/blob/9d541e5c3a107b006cd61e7ece88df5cda7072e0/cli/src/main/scala/scala/scalanative/cli/options/NativeConfigOptions.scala#L40

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: Can you explain how `scala.deriving.Mirror.Singleton` works? Do we need to use it here?

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: Why was this code removed?

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: I think that instead of executing `detectNativePackageType` here, it can be handled in this [section](https://github.com/VirtusLab/scala-cli/blob/d062ea14f3286adbad722117bbac45cb69284085/modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala#L222)

**Review by lwronski — COMMENTED** [review]
It looks promising to me, and I would be happy to merge it. Unfortunately, there is one new test that is failing. You can find more details [here](https://github.com/VirtusLab/scala-cli/actions/runs/5219458233/jobs/9421507356?pr=2196#step:5:2708).

To fix the `checks` and `reference-doc` jobs, run the following commands before pushing your changes:
```
./mill -i __.fix
./mill -i generate-reference-doc.run
```

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (keynmol) [reply modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: Apologies, this is a an artefact of a Metals import - Singleton is not needed it here.

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (keynmol) [reply modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: Human error :)

**Comment by keynmol** [comment]
@lwronski, thanks, that's exactly the commands I was looking for :) 

If you are happy with the overall approach I will start fixing and cleaning up the PR

**Comment by lwronski** [comment]
> @lwronski, thanks, that's exactly the commands I was looking for :)
> 
> If you are happy with the overall approach I will start fixing and cleaning up the PR

Yes, I'm satisfied with the overall approach, so I will wait for the final results.

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (keynmol) [inline modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: @WojciechMazur let me know if you have thoughts here - without it the libraries produced depend on themselves (located in .scala-build) so they're not portable

We're doing "convention over configuration" here for better user experience.

File: modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala
Comment (keynmol) [inline modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala]: Flagging here - I _think_ this was a bug (embedResources not being passed)

**Comment by Gedochao** [comment]
@keynmol what is the current status on this? can we unblock you somehow?

**Comment by keynmol** [comment]
Hey @Gedochao! I didn't forget about this PR, just was short on time.

I resolved the conflicts (but incorrectly it seems, I will address those failures).

~~Now that I re-ran the tests on Linux, I remember the main problem - this PR relies on passing `-install_name` linker flag, to make sure that the dynamic library loading entries refer to the correct (**relative**) location of the built dynamic library, instead of the location where it's actually built (`.scala-build/.../`).~~

~~Problem is - on Linux this flag doesn't seem to work~~

Just as I was writing those last two paragraphs I found the linux equivalent option 🤦 

I'll pick this up again and hopefully will address remaining test failures.

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (keynmol) [inline modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala]: @Gedochao miraculously, the tests are now green.

One issue I discovered though is that I can't seem to override this directive by passing `--native-target` in the command line.

This (from my side) would be the last thing I'd want to address with this PR

Comment (Gedochao) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala]: Nice!
I can try looking at it sometime next week, it's not exactly easy to pass it all the way from the command line.
I'll let you know when I do.

File: modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala
Comment (WojciechMazur) [reply modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: I've never actually used this one, but seems to be required. I'm fine with with adding this flag, and wonder should we backport it to the ScalaNative toolchain

Comment (keynmol) [reply modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala]: I think ScalaNative toolchain has same issue:

```
$ otool -L target/scala-3.3.1/libscala-native-seed-project-out.dylib
target/scala-3.3.1/libscala-native-seed-project-out.dylib:
        /private/var/folders/rv/tdh6r2xx2b7by3z590lq8vrm0000gn/T/tmp.iWqfQcKp/scala-native-seed-project/target/scala-3.3.1/libscala-native-seed-project-out.dylib
```

Basically this dylib is not portable - it references itself in the target folder

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (Gedochao) [inline modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: hey, for this to run on the CI we'd have to have the utility available and have the env variable point to it, right?
Am I assuming correctly that this test is being skipped on the CI right now?

Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: Yes, I explicitly ignored this test for now to save you the pain of installing LLVM on CI :D 

Additionally, I'm working on not needing this at all: https://github.com/scala-native/scala-native/pull/3548

Once that's merged and hopefully backported, we can remove the LLVM_BIN check and things should work out of the box

Comment (Gedochao) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: ah, okay.
I was already looking into setting it up on CI ;D
cool, in that case it's okay to skip this test temporarily 👍

Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: For posterity, installing LLVM is not horribly difficult, but it takes some time (though on the scale of how long the tests take.. it's insignificant)

https://github.com/indoorvivants/sn-bindgen/blob/main/build/setup-debian.sh#L5-L9

Using their official apt packages: https://apt.llvm.org/

And the LLVM_BIN value will be `/usr/lib/llvm-14/bin` where 14 is the version you install.

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala]: We can also go through the package type option (I think I had that originally), but it's a bit confusing with --native-target.

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: If you'd prefer - I can set up LLVM in only one of the workflows that runs this test (just need to figure out which one), to make sure we don't ignore any CI tests unnecessarily

File: modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala
Comment (Gedochao) [inline modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala:277]: @keynmol this is what seems to have been missing for the command line override

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (Gedochao) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala]: I fixed the `--native-target` override in the last commit and added a relevant test, should work correctly now

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (Gedochao) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: If you have the time to set up the workflow, go ahead.
Although it's also fine to leave it temporarily inactive, since it seems it won't be necessary anyway once https://github.com/scala-native/scala-native/pull/3548 gets merged. Maybe easier that way for now.

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala]: nice!

File: modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala
Comment (keynmol) [reply modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:533]: > Maybe easier that way for now

Ok, I won't do it then :)

Once we have a timeline/decision on that PR being backported in 0.4.x, I can go ahead and make a separate PR adding LLVM (if necessary).

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/ScalaNative.scala
Comment (MaciejG604) [inline modules/directives/src/main/scala/scala/build/preprocessing/directives/ScalaNative.scala]:
```suggestion
  nativeEmbedResources: Option[Boolean] = None,
  nativeTarget: Option[String] = None,
```

File: modules/cli/src/main/scala/scala/cli/commands/shared/ScalaNativeOptions.scala
Comment (MaciejG604) [inline modules/cli/src/main/scala/scala/cli/commands/shared/ScalaNativeOptions.scala:66]:
```suggestion
  @Tag(tags.should)
  @ValueDescription("app|static|dynamic")
```

**Review by MaciejG604 — APPROVED** [review]
LGTM

**Review by Gedochao — APPROVED** [review]

**Comment by keynmol** [comment]
🎉 

Thanks @Gedochao and @MaciejG604

**Pull request merged**
